### PR TITLE
Added invitation for native software citation

### PR DESCRIPTION
**Pull request by ivalaginja**
# Summary

Improve information on how to cite Sherpa in the CITATION file and with a new method, sherpa.citation, which will display the citation information provided by Zenodo.

# Details

The sherpa.citation method can either query Zenodo to get the data or look up a hard-coded table (stored in the sherpa module) for past releases. The output is of the form

 - what version of Sherpa you are using
 - the Zenodo citation or a message saying it couldn't be found (the version isn't known, the Zenodo API has changed and no-longer returns the information we are expecting, there is no internet connection, or a glitch in the download)
 - information on how to cite vresions, linking to the Zenodo help page
 - Two "general purpose" references to Sherpa

# Example

If you were using a released version then the citation method would return something like

```
>>> sherpa.citation()
You are using Sherpa 4.12.1.

Sherpa 4.12.1 was released on July 14, 2020.

@software{sherpa_2020_3944985,
  author       = {DougBurke and
                  Omar Laurino and
                  wmclaugh and
                  dtnguyen2 and
                  Marie-Terrell and
                  Hans Moritz Günther and
                  Jamie Budynkiewicz and
                  Aneta Siemiginowska and
                  Tom Aldcroft and
                  Christoph Deil and
                  Brigitta Sipőcz and
                  Katrin Leinweber and
                  Todd},
  title        = {sherpa/sherpa: Sherpa 4.12.1},
  month        = jul,
  year         = 2020,
  publisher    = {Zenodo},
  version      = {4.12.1},
```

If you are using a random commit then you'll see something like the following, except that the response depends on whether the version number exists on Zenodo (in which case you will get that information - ie the `@citation{sherpa_...}` section from above), or a warning that the version isn't known, such as

```
>>> sherpa.citation()
You are using Sherpa 4.13.0+442.g38fb7ea5 (it is not a released version).

There was a problem retireving the data from Zenodo:
Zenodo has no information for version 4.13.0.

Please review the Zenodo Sherpa page at
       https://doi.org/10.5281/zenodo.593753
to identify the latest release of Sherpa. The Zenodo page
https://help.zenodo.org/#versioning provides information on how to
cite a specific version.

If you want a general-purpose citation then please use either of the
following, kindly provided by the SAO/NASA Astrophysics Data System
service:

@INPROCEEDINGS{2001SPIE.4477...76F,
       author = {{Freeman}, Peter and {Doe}, Stephen and {Siemiginowska}, Aneta},
...
```

# Background

Needing to cite Sherpa in a publication that will hopefully happen soon, I checked the `CITATION` file and realized that there is nothing present for a native software citation.
Trying to establish citation principles that will give everyone due credit, as well as create traceable records of a work unit like a piece of software, the [FORCE11 Software Citation Working Group](https://www.force11.org/about) published a [paper presenting software citation principles](https://www.force11.org/software-citation-principles). This includes the need to cite the software product directly, *additionally* to software papers (see section 6.2):
> 6.2 **Software papers.** Currently, and for the foreseeable future, software papers are being published and cited, in addition to software itself being published and cited, as many community norms and practices are oriented towards citation of papers. [...] **the software itself should be cited on the same basis as any other research product; authors should cite the appropriate set of software products**. If a software paper exists and it contains results (performance, validation, etc.) that are important to the work, then the software paper should also be cited. We believe that a request from the software authors to cite a paper should typically be respected, and the paper cited **in addition** to the software.

This prompted me to submit this PR adding a section to the `CITATION` file that asks to cite the software directly as well, which is easy since `sherpa` has a Zenodo DOI and is hence easily citable.

It would be great to see this included and move towards this as a general standard in citing software. Please do advise though in case this is not wanted or wanted to be done differently.

### CODE REVIEW

**Comment by DougBurke** [comment]
So, we would have to make sure to update this for every release (in the same way we update `README.md`) to refer to the latest version (essentially the contents of the bibtex file). Two thoughts:

- can we automatically generate this fragment (i.e. have some form of a pre-processor that grabs the data from Zenodo to create the bibtex)
- or should we just point people to the zenodo site since it has a `Cite As`/`Export` section (and I think we can use the "generic version" link here)

The second option makes it more work for the user, but makes it much-less likely we'll have out-of-date information in this file.

**Comment by ivalaginja** [comment]
I think I would be in favor of the second option because as you say, it makes it much less likely to have out-of-date information, which is never good. The BibTeX entry I added is pointing to the concept DOI instead of the version DOI to mitigate out-of-date information already, putting it on the reader to actually go fetch the version DOI from Zenodo `Cite As` section if they're not using the most recent version. This is also why I mentioned that in the file.

In the end, as much as people have to learn how to cite papers properly, they also need to learn how to cite software properly. Until such time, it is probably the best to be as explicit as possible to make it easier for the user.

**Comment by augustfly** [comment]
hey, i work on this subject (software citation and bibtex etc). @DougBurke chatted me in from twitter a week ago. I'm only now replying.

A few things. 

1. Zenodo recently updated their bibtex to tag the type "software" instead of misc and add publisher and version key fields. It should look something like (yes, I fixed it):

```
@software{sherpa,
  author       = {Doug Burke and
                  Omar Laurino and
                  D.T. Nguyen and
                  Jamie Budynkiewicz and
                  Tom Aldcroft and
                  Aneta Siemiginowska and
                  Christoph Deil and
                  Warren McLaughlin and
                  Brigitta Sipocz and
                  Katrin Leinweber},
  title        = {sherpa/sherpa},
  month        = aug,
  year         = 2019,
  publisher    = {Zenodo},
  version      = {4.11.1},
  doi          = {10.5281/zenodo.3358134},
  url          = {https://doi.org/10.5281/zenodo.3358134}
}

```
2. I'm not sure adding the concept DOI to the bibtex is a good idea. Your bibtex snippet is version specific anyway, and if you're going to be editing it (locally on GH) to update the version (in the title and in the version key) then one might as well update the DOI to the version specific one. 
2. If you wanted to give some example bibtex you could replace the authorship with "The Sherpa Developers", simplify the title, remove the version key, and use the concept DOI.   Then you could tell your users to "follow the concept DOI to find the that version/author/DOI information relevant to the code version they used."  This way if they are lazy they are at least citing the lazy thing (a concept DOI sans authors or version) and not some blended thing. 
2. A bibtex example for a concept DOI might look something like this, which I totally made up because  Zenodo is grappling with what the metadata should look like for concept DOIs (there is no real metadata for a concept doi as yet). 

```
@software{sherpa,
  author       = {The Sherpa Development team},
  title        = { Sherpa: a Python package for modeling and fitting data},
  year         = 2015,
  publisher    = {Zenodo},
  doi          = {10.5281/zenodo.593753},
  url          = {https://doi.org/10.5281/zenodo.593753}
}
```

**Comment by ivalaginja** [comment]
This is an awesome way of explaining it, thanks a lot! I am definitely saving this for future reference for myself.

**Comment by DougBurke** [comment]
I've rebased this PR and added 

- a citation method to the sherpa module
- updated CITATION to refer to this new functionality

The problem with mainting a Zenodo bibtex record in the CITATIONFILE is that we have no automatic procedure to make sure it's up-to-date. So I'd rather make the user query Zenodo to get the correct version (it is an extra step, which is a bummer). The citation function is meant to make this easy. Unfortunately it's not immediate as it has to make a call to Zenodo (and note the screen output is paginated, which isn't shown here):

```
>>> import sherpa
>>> sherpa.citation()  # get the latest release
The latest release of Sherpa is sherpa/sherpa: Sherpa 4.12.1
released on July 14, 2020.

@software{sherpa_2020_3944985,
  author       = {Doug Burke and
                  Omar Laurino and
                  wmclaugh and
                  dtnguyen2 and
                  Marie-Terrell and
                  Hans Moritz Günther and
                  Jamie Budynkiewicz and
                  Aneta Siemiginowska and
                  Tom Aldcroft and
                  Christoph Deil and
                  Brigitta Sipőcz and
                  Katrin Leinweber and
                  Todd},
  title        = {sherpa/sherpa: Sherpa 4.12.1},
  month        = jul,
  year         = 2020,
  publisher    = {Zenodo},
  version      = {4.12.1},
  doi          = {10.5281/zenodo.3944985},
  url          = {https://doi.org/10.5281/zenodo.3944985}
}

Please review the Zenodo Sherpa page at
       https://doi.org/10.5281/zenodo.593753
to identify the latest release of Sherpa. The Zenodo page
https://help.zenodo.org/#versioning provides information on how to
cite the specific version.

If you want a general-purpose citation then please either of the
following, kindly provided by the SAO/NASA Astrophysics Data System
service:

@INPROCEEDINGS{2001SPIE.4477...76F,
       author = {{Freeman}, Peter and {Doe}, Stephen and {Siemiginowska}, Aneta},
        title = "{Sherpa: a mission-independent data analysis application}",
     keywords = {Astrophysics},
    booktitle = {Astronomical Data Analysis},
...
... more bibtexey goodness
...
```

You can also request a specific version which can be faster

```
>>> sherpa.citation('4.12.1')
Sherpa 4.12.1 was released on July 14, 2020.

@software{sherpa_2020_3944985,
  author       = {DougBurke and
                  Omar Laurino and
                  wmclaugh and
                  dtnguyen2 and
                  Marie-Terrell and
                  Hans Moritz Günther and
                  Jamie Budynkiewicz and
                  Aneta Siemiginowska and
                  Tom Aldcroft and
                  Christoph Deil and
                  Brigitta Sipőcz and
                  Katrin Leinweber and
                  Todd},
  title        = {sherpa/sherpa: Sherpa 4.12.1},
  month        = jul,
  year         = 2020,
  publisher    = {Zenodo},
  version      = {4.12.1},
  doi          = {10.5281/zenodo.3944985},
  url          = {https://doi.org/10.5281/zenodo.3944985}
}

Please review the Zenodo Sherpa page at
       https://doi.org/10.5281/zenodo.593753
to identify the latest release of Sherpa. The Zenodo page
https://help.zenodo.org/#versioning provides information on how to
cite the specific version.

If you want a general-purpose citation then please either of the
following, kindly provided by the SAO/NASA Astrophysics Data System
service:

@INPROCEEDINGS{2001SPIE.4477...76F,
       author = {{Freeman}, Peter and {Doe}, Stephen and {Siemiginowska}, Aneta},
        title = "{Sherpa: a mission-independent data analysis application}",
     keywords = {Astrophysics},
...
```

**Comment by DougBurke** [comment]
One thing my edit has done is to remove the "concept" DOI from the CITATION document as (assuming Zenodo hasn't made recent improvements on that score) it looks a tricky thing to get right based on @augustfly 's thoughtful write-up.

File: sherpa/__init__.py [hunk 1]
```diff
@@ -63,9 +63,447 @@ def format(self, record):
...
+    ol = 'Omar Laurino'
+    dn = 'dtnguyen2'
+    ta = 'Tom Aldcroft'
+    an = 'Aneta Siemiginowska'
+
+    jb = 'Jamie Budynkiewicz'
+    cd = 'Christoph Deil'
+    bs = 'Brigitta Sipocz'
+
+    wm = 'wmclaugh'
```
Comment (augustfly) [inline sherpa/__init__.py:182]: If you're able to give everyone a variable then you could give most of these folks real names instead of handles, right?

Comment (DougBurke) [reply sherpa/__init__.py:182]: All I'm trying to do here is match the Zenodo output. While I agree that citing 'and Todd' is not the most informative thing to do, I personally feel that that's outside this particular PR. Or my paygrade.

**Review by hamogu — APPROVED** [review]
This is way more work then I think is justified - but now that you've written it, I want to copy and paste it into other projects, too (or release it as a separate python package...).
We need to do a lot of releases to recoup the time investment that @DougBurke did here. Hat off!

A few typos, otherwise I think this is fine to go and thus I'll click "approve" (while there are code changes, this really is about documentation, so my review is at least a good start).

Although I approve, I want to note that I disagree with one choice in this PR. If @DougBurke still prefers the way he did it after reading this, then that's fine with me, too, but I wanted to bring it up. I understand the thought behind the choice not to test the connection to Zenodo every time. On the other hand, not testing means that we are blind to any change on the Zenodo side. If they change their interface today, we could ship code that never works and won't know. In that case, all we have is the fallback to the hardcoded information, which needs to be updated manually for every release. In other words, we would find ourselves in the same situation that we were in before this giant work: Either we remember to update the citation information manually and test zenodo by hand, or we risk shipping stale information. There is just so much that could go wrong with the internet (changing ssh protocols come to mind) or any API change.

Just like before, I don't see an easy way out. I don't want to waste this effort, but I also want it to be useful. So, I suggest to add a test for a Zenodo query and mark is with `@pytest.mark.skip` or `@pytest.mark.xfail`. In either case, someone will hopefully notice the output every now  and then and manually check why the test was marked that way.

The full solution probably involves a new test marker (`@sherpa.mark.monthly`) to mark tests that are not run on every CI, but just as a monthly cron job on Travis - but setting up all that is too much work for just this one test.

**Comment by DougBurke** [comment]
@hamogu - some concerns I have about running tests against Zenodo are

- it wastes their resources
- technically our tests could run often that we run out of API requests as they are rate limited (I've no idea what address all the Travis calls will come from, so it may not be a real issue)
- I'm not sure I believe we would notice an optional test (so we don't get bitten by transient internet tomfoolery) failing or suddenly passing but perhaps I'm missing the point (I am tired)

I should point out that I'm riffing on the R package citation() method, but they don't have to bother with my over-engineered download-a-palooza code!

File: CITATION [hunk 2]
```diff
@@ -1,38 +1,47 @@
...
 
-BibTeX entries for LaTeX users are
+This will only work if you have internet access (at least for the
+latest query) as it takes the data from the Sherpa record at
+Zenodo: https://zenodo.org/record/3944985
+
+The Zenodo guide to versioning at http://help.zenodo.org/#versioning
+provides advice on how to cite the correct version.
+
+If you want a general-purpose citation then please either of the
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply CITATION]: Ta

**Comment by DougBurke** [comment]
@pkgw - here is where I've ended up with our recent "citation" discussion.

**Comment by pkgw** [comment]
@DougBurke Thanks for the ping! Cool approach.

One thing that struck me is that I'd have expected the argument-less `citation()` command to return the citation for the *currently installed* version, not the *latest* version (if I'm reading correctly). If you use Sherpa v3.0.0 to do your analysis even though the latest version is v24.0.0, it seems to me that the `citation()` should still show you the info for 3.0.0. (Relatedly, the code here would be a good basis for reminding people to upgrade ...)

Overall, somewhat echoing @hamogu's comments, the amount of code needed here really reinforces my wish that there was some kind of "how-do-i-cite.org" service that you could point people to, rather than having to implement all of the logic in-app. If Zenodo's API ever changes (very unlikely, I think, but still), old installs will be stuck and the functionality will just break for them. If you could open a browser window pointing to `how-do-i-cite.org/sherpa?installedversion=4.2.10` then you could preserve the flexibility to arbitrarily "patch" citation instructions for old installs as needed.

**Comment by DougBurke** [comment]
@pkgw if onli I knew an innovation fellow I could hoist this project off on ;-)

**Comment by DougBurke** [comment]
@hamogu - I've added a `--runzenodo` test option that will cause two tests to run: a check that we can call Zenodo, and a check that it errors out if the version number is unknown. You have to select this option, so it won't be run in our Travis tests.

**Comment by pkgw** [comment]
CC @dbouquin if you've got any thoughts about the approach here?

**Comment by DougBurke** [comment]
@pkgw - I've just pushed a change to follow your suggestion by matching the release by default. This means that users of the development versions will get less-than-ideal responses: e.g. at the moment you get

```
>>> sherpa.citation()
You are using Sherpa 4.13.0+442.g38fb7ea5 (it is not a released version).

There was a problem retireving the data from Zenodo:
Zenodo has no information for version 4.13.0.

Please review the Zenodo Sherpa page at
       https://doi.org/10.5281/zenodo.593753
to identify the latest release of Sherpa. The Zenodo page
https://help.zenodo.org/#versioning provides information on how to
cite a specific version.

If you want a general-purpose citation then please use either of the
following, kindly provided by the SAO/NASA Astrophysics Data System
service:

@INPROCEEDINGS{2001SPIE.4477...76F,
       author = {{Freeman}, Peter and {Doe}, Stephen and {Siemiginowska}, Aneta},
...
```

but

a) I've added the current version number at the start, and an indication it is not a released version
b) I don't expect most science will be done with a development version, but a released version

For  reference, if you were using a release version then you would see something like

```
>>> sherpa.citation()
You are using Sherpa 4.12.1.

Sherpa 4.12.1 was released on July 14, 2020.

@software{sherpa_2020_3944985,
  author       = {DougBurke and
                  Omar Laurino and
                  wmclaugh and
                  dtnguyen2 and
                  Marie-Terrell and
                  Hans Moritz Günther and
                  Jamie Budynkiewicz and
                  Aneta Siemiginowska and
                  Tom Aldcroft and
                  Christoph Deil and
                  Brigitta Sipőcz and
                  Katrin Leinweber and
                  Todd},
  title        = {sherpa/sherpa: Sherpa 4.12.1},
  month        = jul,
  year         = 2020,
  publisher    = {Zenodo},
  version      = {4.12.1},
...
```

**Comment by hamogu** [comment]
Great! No more comments from me. Let's get this in!

**Comment by DougBurke** [comment]
Final round of testing, where I was manually inserting failures from the Zenodo query to mimic the network being down or the response changing pointed out a problem which I've now fixed.

I hope it is possible to insert some machinery to allow the tests to check these conditions (by using mock objects to insert invalid data), but I don't know how to do this at this time, and it would take too ling to investigate, hence the manual checks.

**Review by dtnguyen2 — APPROVED** [review]

**Comment by wmclaugh** [comment]
@DougBurke - Can you please rebase this PR

**Comment by DougBurke** [comment]
*argh*

**Comment by DougBurke** [comment]
@wmclaugh - hopefully this works. I'll check the travis output tomorrow as it's nighty-night time for me!

**Comment by DougBurke** [comment]
@ivalaginja - thanks for starting this, and apologies that it took so long to address.

**Comment by ivalaginja** [comment]
@DougBurke there are certainly no apologies needed, I was curiously following how you guys were coming up with new solutions to address this - kudos! Fantastic to see this happen, and thanks for picking up my suggestion!

**Pull request merged**